piqueserver/config.py

```diff
@@ -0,0 +1,145 @@
+# Copyright (c) 2017 Piqueserver development team
+
+# This file is part of piqueserver.
+
+# piqueserver is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# piqueserver is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with piqueserver.  If not, see <http://www.gnu.org/licenses/>.
+
+import json
+import toml
+
+
+DEFAULT_STYLE = 'TOML'
+TOML_STYLE = 'TOML'
+JSON_STYLE = 'JSON'
+
+
+class ConfigException(Exception):
+    pass
+
+
+class ConfigStore():
+    '''
+    configuration store that manages global configuration
+    '''
+
+    def __init__(self):
+        self._raw_config = {}
+        self.options = {}
+
+    def validate_all(self):
+        for option in self.options.values():
+            option.validate(option.get())
+
+    def clear_config(self):
+        self._raw_config = {}
+
+    def load_config(self, config_file, style=DEFAULT_STYLE):
+        self.clear_config()
+        self.update_config(config_file, style)
+
+    def update_config(self, config_file, style=DEFAULT_STYLE):
+        if style == TOML_STYLE:
+            self._raw_config.update(toml.load(open(config_file)))
+        elif style == JSON_STYLE:
+            self._raw_config.update(json.load(open(config_file)))
+        else:
+            raise ConfigException('Unsupported config file format: {}'.format(style))
+        self.validate_all()
+
+    def load_config_object(self, config):
+        self._raw_config = config
+        self.validate_all()
+
+    def update_config_object(self, config):
+        self._raw_config.update(config)
+        self.validate_all()
+
+    def dump_config(self, out_file, style=DEFAULT_STYLE):
+        if style == TOML_STYLE:
+            toml.dump(self._raw_config, open(out_file, 'w'))
+        elif style == JSON_STYLE:
+            json.dump(self._raw_config, open(out_file, 'w'))
+        else:
+            raise ConfigException('Unsupported config file format: {}'.format(style))
+
+    def get_raw_config(self):
```

Comment:
I don't think we should have this. If people *really* want to access the internals, they can do it via the _raw_config attribute and will also be warned by the underscore that they are doing stupid shit and should probably overthing it.

Reply:
I was thinking it could be useful to support backwards compatibility with scripts. What do you think?